### PR TITLE
[DSIP-89][SqlTask] Avoid using PrepareStatement in SqlTask

### LINKED ISSUE #17263: [DSIP-89][SqlTask] Avoid using PrepareStatement in SqlTask
### Search before asking

- [x] I had searched in the [DSIP](https://github.com/apache/dolphinscheduler/issues/14102) and found no similar DSIP.

### Motivation

Right now we're using PrepareStatement in SqlTask, and if we find the sql contains `${xx}` will using `?` to replase `${xx}` and set the value in `PrepareStatement`. 

There might exist a lot of bug in this way, some are from DS, some are from third-part driver.
e.g. #17124

So it's better to directly replace the value in sql rather than using Preparement.

### Design Detail

_No response_

### Compatibility, Deprecation, and Migration Plan

_No response_

### Test Plan

_No response_

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qiuyanjun888**
## Was this PR generated or assisted by AI?

YES. This pull request was assisted by Hermes Agent / OpenAI Codex for code changes, focused tests, documentation updates, and local verification. The scope and final submission were directed by the contributor.

## Purpose of the pull request

Fix #17263.

Render SQL task custom parameters locally before submitting SQL through JDBC `Statement`, avoiding `PreparedStatement` for SQL text that already contains DolphinScheduler parameter syntax.

This pull request supersedes #18396, which was closed because the original description did not include the AI-assistance declaration required by the current pull request template.

## Brief change log

- Add a SQL task local parameter renderer for `${name}` and `!{name}` placeholders.
- Render normal `${name}` placeholders as SQL literals by default, including escaping string-like values and validating typed literals.
- Preserve raw `!{name}` replacement for trusted SQL fragments without rescanning rendered replacement text.
- Use JDBC `Statement` to execute rendered SQL in `SqlTask`.
- Document SQL task parameter rendering rules and raw replacement constraints.
- Add regression coverage for identifier fragments, list values, quoted strings, raw fragments, malformed typed literals, and Statement execution.

## Verify this pull request

This change added tests and can be verified as follows:

- `./mvnw -pl dolphinscheduler-task-plugin/dolphinscheduler-task-sql -am -DskipUT=false -DskipITs -DskipE2E -Dcheckstyle.skip -Dspotless.check.skip -Djacoco.skip=true -Dsurefire.failIfNoSpecifiedTests=false -Dtest=SqlTaskTest test`
- `./mvnw -pl dolphinscheduler-task-plugin/dolphinscheduler-task-sql -am spotless:apply`
- `git diff --check`

## Pull Request Notice

[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

### CODE REVIEW

File: dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTaskParameterRenderer.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        String value = property.getValue();
+        if (value == null) {
+            return "null";
+        }
+        if (insideStringLiteral) {
+            return escapeSqlString(value);
+        }
+        if (identifierContext) {
+            return value;
+        }
```
[the remark this answers is not in the document]
Comment (qiuyanjun888) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/main/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTaskParameterRenderer.java]: Fixed in `0ca7e2e` by validating `${...}` values in identifier-fragment context against a non-empty allowlist of letters, digits, underscores, and dollar signs. Invalid fragments now throw `TaskException` before SQL execution; raw `!{...}` replacement remains unchanged.

Verification: targeted `SqlTaskTest` passed (29 tests), `spotless:check` passed, and `git diff --check` passed.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/test/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTaskTest.java [hunk 2]
```diff
@@ -479,6 +484,233 @@ void testEnsureSqlContent_whenResourceMissing_throwsTaskException(@TempDir Path
...
+        method.setAccessible(true);
+
+        SqlBinds binds = (SqlBinds) method.invoke(task,
+                "select * from student where name=${name}");
+
+        Assertions.assertEquals("select * from student where name='!{fragment}'", binds.getSql());
+    }
+
+    @Test
+    void testSqlTaskLocalRenderer_keepsPlaceholderInsideJsonStringLiteral() throws Exception {
```
[the remark this answers is not in the document]
Comment (qiuyanjun888) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-sql/src/test/java/org/apache/dolphinscheduler/plugin/task/sql/SqlTaskTest.java]: Fixed in 87add3589c0a9cae9ba43cf794aabeb9b94d2b70: renamed the test to `testSqlTaskLocalRenderer_replacesPlaceholderInsideJsonStringLiteral`, matching the assertion that `${name}` is rendered inside the JSON string literal.

Validation:
- Targeted test: 1 run, 0 failures/errors/skips
- `./mvnw -pl dolphinscheduler-task-plugin/dolphinscheduler-task-sql spotless:check`: passed

**Comment by SbloodyS** [comment]
The local string escaping is not safe across the supported database dialects.

`quoteSqlString()` only doubles single quotes. In MySQL's default mode, backslash is also an escape character. For example, a value such as:

```text
\'; DROP TABLE audit_log; -- 
```


```markdown
`render()` returns immediately when `paramsMap` is null or empty. Consequently, SQL such as:

```sql
select * from t where id = ${missing}

**Comment by qiuyanjun888** [comment]
> The local string escaping is not safe across the supported database dialects.
> 
> `quoteSqlString()` only doubles single quotes. In MySQL's default mode, backslash is also an escape character. For example, a value such as:
> 
> ```
> \'; DROP TABLE audit_log; -- 
> ```
> 
> ```
> `render()` returns immediately when `paramsMap` is null or empty. Consequently, SQL such as:
> 
> ```sql
> select * from t where id = ${missing}
> ```


Removed the dialect-independent automatic quoting, escaping, type conversion, list rendering, and identifier-context inference. Both ${...} and !{...} now perform a single-pass textual replacement, so datasource-specific quoting remains in the SQL or parameter value.
Removed the null/empty parameter-map bypass. Missing parameters and unclosed source placeholders now fail with TaskException before the SQL reaches JDBC execution.
Kept replacement values non-recursive: placeholder-like text introduced by a value is not scanned again.
Updated the English and Chinese SQL Task documentation with the textual-replacement and trusted-input contract.
Commit: https://github.com/apache/dolphinscheduler/commit/de866c7d7c3d17ec850cf04c687351204a53223c

Validation: SqlTaskTest — 31 tests, 0 failures/errors/skips; targeted Spotless check passed.

**Review by SbloodyS — COMMENTED** [review]
Thanks for the update. The previous escaping and missing-parameter concerns were addressed, but I still see two blocking compatibility regressions.

### 1. [P1] Preserve the existing `${...}` value semantics

`SqlTaskParameterRenderer.renderParameters()` now appends `Property.getValue()` verbatim and ignores its `DataType`. This breaks existing SQL Task parameters that previously relied on JDBC binding:

- `where name=${name}` with `name=alice` becomes `where name=alice`, treating the value as an identifier.
- `where name='${name}'` with `name=O'Reilly` produces invalid SQL.
- A `LIST` parameter is normally represented as JSON. For example, `[1,2]` previously expanded to two bound parameters, but now `IN (${ids})` becomes `IN ([1,2])`.

The new list test avoids this regression by manually storing `1,'x',3` in a `LIST` property, which does not match the existing runtime representation used by `ParameterUtils.expandListParameter()`.

Since `!{...}` already provides raw SQL-fragment replacement, `${...}` should retain value/type semantics, or this must be treated as an explicitly approved incompatible change with a complete migration strategy and realistic regression tests for VARCHAR, DATE, BOOLEAN, and LIST parameters.

### 2. [P1] Do not reject datasource-native placeholders

Every source `${...}` expression is now treated as a DolphinScheduler custom parameter, and `getProperty()` throws when it is absent from `prepareParamsMap`.

Consequently, valid datasource syntax such as `${hiveconf:dt}` fails before reaching Hive when the task has no corresponding DolphinScheduler parameter. Previously, SQL was passed through unchanged when the parameter map was null or empty.

The existing non-rescanning test only covers `${hiveconf:dt}` introduced by a replacement value; it does not cover the placeholder appearing directly in the source SQL. Please distinguish DolphinScheduler parameters from datasource-native placeholders—or preserve unknown placeholders—and add a regression test using source SQL with `${hiveconf:dt}` and an empty parameter map.

This PR also changes established SQL Task behavior but does not document the migration in `docs/docs/en/guide/upgrade/incompatible.md`, as required by the PR template.
